source/common/router/router_ratelimit.cc

```diff
@@ -63,12 +70,30 @@ bool RemoteAddressAction::populateDescriptor(
 
 bool GenericKeyAction::populateDescriptor(const Router::RouteEntry&,
                                           RateLimit::Descriptor& descriptor, const std::string&,
-                                          const Http::HeaderMap&,
-                                          const Network::Address::Instance&) const {
+                                          const Http::HeaderMap&, const Network::Address::Instance&,
+                                          const envoy::config::core::v3::Metadata*) const {
   descriptor.entries_.push_back({"generic_key", descriptor_value_});
   return true;
 }
 
+DynamicMetaDataAction::DynamicMetaDataAction(
+    const envoy::config::route::v3::RateLimit::Action::DynamicMetaData& action)
+    : metadata_key_(action.metadata_key()), descriptor_key_(action.descriptor_key()) {}
+
+bool DynamicMetaDataAction::populateDescriptor(
+    const Router::RouteEntry&, RateLimit::Descriptor& descriptor, const std::string&,
+    const Http::HeaderMap&, const Network::Address::Instance&,
+    const envoy::config::core::v3::Metadata* dynamic_metadata) const {
+  const ProtobufWkt::Value& metadata_value =
+      Envoy::Config::Metadata::metadataValue(dynamic_metadata, metadata_key_);
+  if (metadata_value.kind_case() != ProtobufWkt::Value::kStringValue) {
```

Comment:
In the docs for the descriptor type can you document this requirement?

Reply:
Done.